### PR TITLE
feat: do constant folding and timezone transformation before create funtional indexes

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Resolve https://github.com/risingwavelabs/risingwave/issues/14762
- Since our batch queries will do constant folding and timezone transformation, functional indexes should store expression in the same form to make index selection happy.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/frontend/src/optimizer/mod.rs [hunk 1]
```diff
@@ -250,6 +235,19 @@ impl PlanRoot {
...
+        if ctx.is_explain_trace() {
+            ctx.trace("To Batch Plan:");
+            ctx.trace(plan.explain_to_string());
+        }
+
+        plan = plan.optimize_by_rules(&OptimizationStage::new(
+            "Merge BatchProject",
+            vec![BatchProjectMergeRule::create()],
+            ApplyOrder::BottomUp,
+        ));
```
Comment (fuyufjh) [inline src/frontend/src/optimizer/mod.rs:249]: Shall we do this before "Convert to physical plan node"?

File: src/frontend/planner_test/tests/testdata/output/index_selection.yaml [hunk 2]
```diff
@@ -740,3 +740,20 @@
   batch_plan: |-
     BatchExchange { order: [], dist: Single }
     └─BatchScan { table: people_names, columns: [people_names.first_name, people_names.last_name, people_names.info], scan_ranges: [people_names.CONCAT_OP = Utf8("John Smith")], distribution: SomeShard }
+- name: funtional index const eval
+  sql: |
+    create table a(v int);
+    create index b on a((1+1)*v);
```
Comment (fuyufjh) [inline src/frontend/planner_test/tests/testdata/output/index_selection.yaml:746]: Is it possible to check the plan of `create index` in the planner test case as well?

**Review by fuyufjh — APPROVED** [review]
I thought we have alread got rid of `AtTimeZone` but obviously it's not? 🤣 PTAL @xiangjinwu

**Review by st1page — APPROVED** [review]

**Comment by xiangjinwu** [comment]
> I thought we have already got rid of `AtTimeZone` but obviously it's not? 🤣 PTAL @xiangjinwu

* `AtTimeZone` is a PostgreSQL user-facing expression that will always be there.
* Our inlining with `AtTimeZone` will be phased out (progress: #14313). But as a safe option I am proposing to disable it with a session variable first before completely removing it.

File: src/frontend/src/optimizer/mod.rs (hunk 1 above)
Comment (chenzl25) [reply src/frontend/src/optimizer/mod.rs:249]: This rule is necessary, because `Convert to physical plan node` might generate additional project.

**Pull request merged**